### PR TITLE
[Store] add remote tensor batch interfaces

### PR DESCRIPTION
**Pull request by zxpdemonio**
## Motivation                                                                                                                                                                      
                                                                                                                                                                                     
  RL training pipelines frequently exchange large DataProto-style batches between rollout, logprob, reward, critic, and trainer stages. These batches contain dense tensor fields,   
  non-tensor rollout data, and lightweight metadata. Sending the whole object through a general-purpose object store can force Python serialization/deserialization of large payloads
   and dominate end-to-end latency.                                                                                                                                                  
                                                                                                                                                                                     
  This PR moves the generic transfer pieces into Mooncake so downstream RL frameworks can keep their own batch semantics locally while using Mooncake for remote tensor/DataProto    
  storage, selected field materialization, and reusable registered buffers.                                                                                                          
                                                                                                                                                                                     
  ## Scenario                                                                                                                                                                        
                                                                                                                                                                                     
  This targets remote DataProto-style transfer in RL pipelines:                                                                                                                      
                                                                                                                                                                                     
  1. A producer stores dense tensor fields as Mooncake objects and records framework-neutral metadata for field names, shapes, dtypes, and lazy selections.                          
  2. Non-tensor batch fields and `meta_info` are transferred through the DataProto helper without requiring a framework-specific schema in Mooncake.                                 
  3. A consumer materializes the full object or selected DataProto field groups through Mooncake.                                                                                    
  4. Tensor materialization can reuse registered local buffers so repeated reads avoid registration on the hot path.                                                                 
                                                                                                                                                                                     
  The helper layer is framework-neutral: ROLL, slime, or another RL framework only needs to provide a compatible container contract instead of changing its rollout schema.          
                                                                                                                                                                                     
  ## Description                                                                                                                                                                     
                                                                                                                                                                                     
  This PR adds and simplifies three groups of Python-facing APIs.                                                                                                                    
                                                                                                                                                                                     
  ### Store APIs                                                                                                                                                                     
                                                                                                                                                                                     
  - `get_tensor_dim_selection_into(key, buffer_ptr, size, shape, dtype, dim, selections=None, data_offset=0)`                                                                        
    - materializes selected slices from one fixed-shape tensor object into a registered output buffer;                                                                               
    - supports arbitrary single-dimension selection through explicit `dim`;                                                                                                          
    - supports negative dim normalization;                                                                                                                                           
    - supports raw tensor payload output without extra TensorMetadata in the destination buffer.                                                                                     
  - `put_from(key, buffer_ptr, size, config=ReplicateConfig{})`                                                                                                                      
  - `batch_put_from(keys, buffer_ptrs, sizes, config=ReplicateConfig{})`                                                                                                             
    - publish raw object payload buffers with explicit `ReplicateConfig`.                                                                                                            
  - `pub_tensor_from(key, buffer_ptr, size, config=ReplicateConfig{})`                                                                                                               
  - `batch_pub_tensor_from(keys, buffer_ptrs, sizes, config=ReplicateConfig{})`                                                                                                      
    - publish prebuilt serialized tensor buffers whose layout is `[TensorMetadata][tensor data]`.                                                                                    
  - `RegisteredBufferPool` / `RegisteredBufferLease` C++ bindings                                                                                                                    
    - manage reusable Mooncake-registered scratch buffers;                                                                                                                           
    - support acquire/release, prewarm, stats, close, and external registration leases.                                                                                          
                                            
  ### Framework-neutral DataProto transfer                                                                                                                                           
                                                                                                                                                                                     
  - `MooncakeDataProtoTransferBackend`                                                                                                                                               
  - `RemoteDataProtoRef`                                                                                                                                                             
  - `RemoteDataProtoShardedRef`                                                                                                                                                      
  - `DataProtoShardPolicy`                                                                                                                                                           
  - `DataProtoMaterializePolicy`                                                                                                                                                     
  - `put_dataproto()` / `materialize_dataproto()` / `remove_dataproto()`                                                                                                             
                                                                                                                                                                                     
  The DataProto helper transfers the full structured object contract:                                                                                                                
                                                                                                                                                                                     
  - `batch`: dense tensors stored as Mooncake objects;                                                                                                                               
  - `non_tensor_batch`: object arrays encoded leaf-by-leaf with specialized codecs and trusted pickle fallback;                                                                      
  - `meta_info`: metadata stored separately and restored during materialization.                                                                                                     
                                                                                                                                                                                     
  It also supports row-sharded references and selected materialization by:                                                                                                           
                                                                                                                                                                                     
  - `batch_fields`                                                                                                                                                               
  - `non_tensor_fields`                       
  - `meta_info_keys`                                                                                                                                                                 
  - `include_meta_info`                                                                                                                                                              
                                                                                                                                                                                     
  Row-level selected DataProto materialization through `row_slice` / `indices` is not implemented yet and currently raises `NotImplementedError`.                                    
                                                                                                                                                                                     
  ### Remote tensor batch and buffer pool helpers                                                                                                                                    
                                                                                                                                                                                     
  - `TensorFieldRef`                                                                                                                                                                 
  - `TensorDimSelection`                                                                                                                                                             
  - `TensorReadRequest`                                                                                                                                                              
  - `RemoteTensorBatch`                                                                                                                                                              
  - `RegisteredBufferPool` in `mooncake.registered_buffer_pool`                                                                                                                  
                                     
  `remote_tensor_batch.py` is now intentionally metadata-only. The reusable buffer lifecycle moved out of that file into `registered_buffer_pool.py`, backed by C++ bindings when
  available and a Python fallback for tests or unrecompiled local extensions.
                                           
  Supported tensor selection ops:                                                                                                                                                    
                                                                                                                                                                                     
  - `select_idxs`                                                                                                                                                                    
  - `slice`                                                                                                                                                                          
  - `repeat`                                                                                                                                                                         
  - `cat`                                                                                                                                                                            
                                                                                                                                                                                     
  Supported dtype names include plain names and torch-style names, e.g. `float32` and `torch.float32`.                                                                               
                                                                                                                                                                                 
  ## Basic Usage                                                                                                                                                                     
                                                                                                                                                                                     
  Remote tensor metadata:                                                                                                                                                            
                                                                                                                                                                                     
  ```python                                                                                                                                                                          
  remote = RemoteTensorBatch(                                                                                                                                                        
      fields={                                                                                                                                                                       
          "input_ids": TensorFieldRef(                                                                                                                                           
              key="batch/input_ids",
              shape=(batch_size, seq_len),
              dtype="int64",               
          )                                
      },                                    
      batch_size=batch_size,      
  )                              
                                     
  selected = remote.select_indices(row_ids)
  requests = selected.read_requests(["input_ids"])

  DataProto transfer:                         
                                          
  backend = MooncakeDataProtoTransferBackend(                                                                                                                                        
      store,                                                                                                                                                                         
      key_prefix="rl/dataproto",                                                                                                                                                     
      data_cls=MyStructuredData,                                                                                                                                                     
  )                                                                                                                                                                              
                             
  ref = backend.put_dataproto(             
      structured_data,                     
      shard_policy=DataProtoShardPolicy(adaptive=True),
  )                               
                                 
  try:                               
      materialized = backend.materialize_dataproto(
          ref,    
          materialize_policy=DataProtoMaterializePolicy(max_inflight_get=4),
      )                                       
  finally:                                
      backend.remove_dataproto(ref)                                                                                                                                                  
                                                                                                                                                                                     
  Selected DataProto field materialization:                                                                                                                                          
                                                                                                                                                                                     
  materialized = backend.materialize_dataproto(                                                                                                                                  
      ref,                   
      batch_fields=["input_ids", "attention_mask"],
      non_tensor_fields=["sample_indices"],
      meta_info_keys=["global_step"],       
  )                               
                                 
  Reusable registered buffers:       
                                         
  from mooncake.registered_buffer_pool import RegisteredBufferPool
                                           
  object_size = 64 * 1024 * 1024              
                                          
  with RegisteredBufferPool(store, max_bytes=2 * 1024**3) as pool:                                                                                                                   
      with pool.acquire(object_size) as lease:                                                                                                                                       
          store.get_into(key, lease.ptr, object_size)                                                                                                                                
                                                                                                                                                                                     
  Architecture                                                                                                                                                                   
                                 
  The implementation keeps framework semantics outside Mooncake and only adds generic structured-data, tensor-selection, and buffer-reuse primitives.
                                         
  Framework container contract
                │
                ▼                             
  MooncakeDataProtoTransferBackend        
                │                                                                                                                                                                    
                ├── batch tensors ───────────────► Mooncake tensor/payload objects                                                                                                   
                ├── non_tensor_batch leaves ─────► encoded Mooncake objects                                                                                                          
                └── meta_info / manifests ───────► metadata objects                                                                                                                  
                                                                                                                                                                                 
  RemoteTensorBatch / TensorReadRequest
                │            
                ▼                          
  get_tensor_dim_selection_into(shape, dtype, dim, selections)
                │                           
                ▼                 
  Tensor dimension selection planner
                │                    
                ▼                        
  Existing get_into_ranges execution path
                │
                ▼                             
  Registered destination buffer           
                                                                                                                                                                                     
  The tensor-selection planner assumes contiguous row-major tensor layout. For common RL dim-0 row selection, selected rows are large contiguous blocks and adjacent rows can be     
  coalesced. For non-leading dimensions, scattered selections can produce more fragments; callers should prefer contiguous slice selections when possible.                           
                                                                                                                                                                                     
  Robustness and Safety                                                                                                                                                          
                                       
  - Validates destination buffers before transfer for non-empty outputs.
  - Rejects invalid shape, dtype, dim, negative repeat, mismatched bool mask length, oversized outputs, and excessive fragments.
  - Uses overflow-checked size and offset arithmetic.
  - Skips Store calls for zero-byte Python materialization requests.
  - Cleans up partial DataProto writes on local write failures.
  - Keeps reusable registered buffers out of the hot registration path.
  - Requires active leases to be released before pool close.
  - Keeps writable Python buffer exports alive through helper regions and releases temporary views idempotently.
  - Documents pickle fallback as trusted-internal-data only.
                  
  Performance
                                              
  This PR reuses the existing Mooncake Store range transfer path. The expected benefit comes from avoiding general Python object-store serialization of large tensor payloads and
  from reusing registered buffers during materialization.                                                                                                                            
                                                                                                                                                                                     
  Downstream framework benchmarks should report:                                                                                                                                     
                                                                                                                                                                                     
  - serialization/encoding cost separately from Mooncake transfer cost;                                                                                                          
  - registered-buffer setup/prewarm cost separately from online read latency;
  - RDMA results separately from TCP or in-memory smoke tests.
                                 
  No framework-specific benchmark script is included in this PR.
                                         
  Module          

  - Transfer Engine (mooncake-transfer-engine)
  - Mooncake Store (mooncake-store)       
  - Mooncake EP (mooncake-ep)                                                                                                                                                        
  - Integration (mooncake-integration) (Python binding)                                                                                                                              
  - P2P Store (mooncake-p2p-store)                                                                                                                                                   
  - Python Wheel (mooncake-wheel)                                                                                                                                                    
  - PyTorch Backend (mooncake-pg)                                                                                                                                                
  - Mooncake RL (mooncake-rl)
  - CI/CD                                  
  - Docs                                   
  - Other                                   
                                       
  Type of Change                 
                                     
  - Bug fix                                
  - New feature                            
  - Refactor                                
  - Breaking change                           
  - Documentation update                  
  - Other                                                                                                                                                                            
                                                                                                                                                                                     
  How Has This Been Tested?                                                                                                                                                          
                                                                                                                                                                                     
  - ./scripts/code_format.sh --check                                                                                                                                             
  - /snap/bin/ruff format --check mooncake-wheel/mooncake/dataproto_transfer.py mooncake-wheel/mooncake/remote_tensor_batch.py mooncake-wheel/mooncake/registered_buffer_pool.py 
  mooncake-wheel/tests/test_dataproto_transfer.py mooncake-wheel/tests/test_put_get_tensor.py mooncake-wheel/tests/test_registered_buffer_pool.py                                    
  - /snap/bin/ruff check mooncake-wheel/mooncake/dataproto_transfer.py mooncake-wheel/mooncake/remote_tensor_batch.py mooncake-wheel/mooncake/registered_buffer_pool.py              
  mooncake-wheel/tests/test_dataproto_transfer.py mooncake-wheel/tests/test_put_get_tensor.py mooncake-wheel/tests/test_registered_buffer_pool.py                                    
  - PYTHONPATH=/root/Mooncake-ROLL/mooncake-wheel /usr/bin/python3 -m pytest -q mooncake-wheel/tests/test_dataproto_transfer.py mooncake-wheel/tests/test_registered_buffer_pool.py  
  - PYTHONPATH=/root/Mooncake-ROLL/mooncake-wheel /usr/bin/python3 -m pytest -q mooncake-wheel/tests/test_put_get_tensor.py::TestRemoteTensorBatchHelpers                        
                             
  Note: the full test_put_get_tensor.py file also contains live MooncakeDistributedStore.setup() tests. In the current local environment, that distributed-store setup aborts, so the
   committed basic validation only runs the helper tests from that file.
  ```                                                            
                                            
  Checklist                       
                                 
  - I have performed a self-review of my own code.
  - I have formatted my own code using ./scripts/code_format.sh before submitting.
  - I have updated the documentation.      
  - I have added tests to prove my changes are effective.

### CODE REVIEW

File: mooncake-integration/store/store_py_parallel_read.h [hunk 1]
```diff
@@ -1096,6 +1178,475 @@ py::list execute_tensor_into_plans(const std::vector<TensorIntoPlan> &plans) {
...
+                return std::nullopt;
+            }
+            if (current.size() != 0 &&
+                static_cast<size_t>(repeat_times) >
+                    std::numeric_limits<size_t>::max() / current.size()) {
+                LOG(ERROR) << context << ": repeat size overflow";
+                return std::nullopt;
+            }
+            size_t repeated_size =
+                current.size() * static_cast<size_t>(repeat_times);
```
Comment (ShangmingCai) [inline mooncake-integration/store/store_py_parallel_read.h]: Should we check if this could potentially cause a size_t overflow?

**Comment by zxpdemonio** [comment]
Some sub-features of this pr are splitted into:
https://github.com/kvcache-ai/Mooncake/pull/2095
https://github.com/kvcache-ai/Mooncake/pull/2140
https://github.com/kvcache-ai/Mooncake/pull/2600
https://github.com/kvcache-ai/Mooncake/pull/2521
https://github.com/kvcache-ai/Mooncake/pull/2472

**Review by stmatengss — APPROVED** [review]
